libraries/chain/fork_database.cpp

```diff
@@ -21,79 +20,117 @@ namespace eosio::chain {
    }
 
-   template<class bsp>  // either [block_state_legacy_ptr, block_state_ptr], same with block_header_state_ptr
+   // match comparison of by_best_branch
+   template<class BS>
+   bool first_preferred( const BS& lhs, const BS& rhs ) {
+      using BSA = BS::fork_db_block_state_accessor;
+      return std::make_tuple(BSA::last_final_block_num(lhs), BSA::lastest_qc_claim_block_num(lhs), BSA::block_height(lhs)) >
+             std::make_tuple(BSA::last_final_block_num(rhs), BSA::lastest_qc_claim_block_num(rhs), BSA::block_height(rhs));
+   }
+
+   template<class BSP>  // either [block_state_legacy_ptr, block_state_ptr], same with block_header_state_ptr
    struct fork_database_impl {
-      using bs               = bsp::element_type;
-      using bhsp             = bs::bhsp_t;
-      using bhs              = bhsp::element_type;
-      
-      using fork_db_t        = fork_database_t<bsp>;
+      using BS               = BSP::element_type;
+      using BSAccessor       = BS::fork_db_block_state_accessor;
+      using BHSP             = BS::bhsp_t;
+      using BHS              = BHSP::element_type;
+
+      using fork_db_t        = fork_database_t<BSP>;
       using branch_type      = fork_db_t::branch_type;
       using full_branch_type = fork_db_t::full_branch_type;
       using branch_type_pair = fork_db_t::branch_type_pair;
 
       using fork_multi_index_type = multi_index_container<
-         bsp,
+         BSP,
          indexed_by<
-            hashed_unique<tag<by_block_id>, BOOST_MULTI_INDEX_CONST_MEM_FUN(bs, const block_id_type&, id), std::hash<block_id_type>>,
-            ordered_non_unique<tag<by_prev>, const_mem_fun<bs, const block_id_type&, &bs::previous>>,
-            ordered_unique<tag<by_lib_block_num>,
-                           composite_key<bs, BOOST_MULTI_INDEX_CONST_MEM_FUN(bs, bool, is_valid),
-                                         // see first_preferred comment
-                                         BOOST_MULTI_INDEX_CONST_MEM_FUN(bs, uint32_t, irreversible_blocknum),
-                                         BOOST_MULTI_INDEX_CONST_MEM_FUN(bs, uint32_t, block_num),
-                                         BOOST_MULTI_INDEX_CONST_MEM_FUN(bs, const block_id_type&, id)>,
-                           composite_key_compare<std::greater<bool>, std::greater<uint32_t>, std::greater<uint32_t>, sha256_less>>>>;
+            hashed_unique<tag<by_block_id>, BOOST_MULTI_INDEX_CONST_MEM_FUN(BS, const block_id_type&, id), std::hash<block_id_type>>,
+            ordered_non_unique<tag<by_prev>, const_mem_fun<BS, const block_id_type&, &BS::previous>>,
+            ordered_unique<tag<by_best_branch>,
+               composite_key<BS,
+                  global_fun<const BS&,             bool,     &BSAccessor::is_valid>,
```

Comment:
Why not use ` const_mem_fun<BS,  bool, &BS::is_valid>` instead of adding a member in the BSAccessors?

Same for `set_valid`, there was already a function available in both block_state versions.

Reply:
So as a result the `BSAccessor` would be needed only for the index definition.

Using the accessor is not optimal, because for the legacy version we compare twice on the same value `bs.irreversible_blocknum();`.

I think it would be better to have two separate definitions for the `by_best_branch` index, which would directly access their respective `block_state` and `block_state_legacy` members without accessors.

Reply:
The BSAccessor for valid is used because I made it private. It really should be private as it is modified after other threads have access to the `block_state`.

I think to have two diff definitions of `by_best_branch` we would have to make `fork_multi_index_type` a template parameter of `fork_database_impl`. I'm not necessarily opposed to that, but I doubt the extra comparison will make much difference. If the BSAccessor and extra comparison is worth the effort, I can make the change. What do you think?

Reply:
I agree that the extra comparison is not a big issue. 
 I was just thinking that the resulting code with my suggestion would be simpler and cleaner (and I don't believe it would require making fork_multi_index_type a template parameter of fork_database_impl). Let me play with it to see if I can make it simpler and we can decide which way is the nicest.

Reply:
see [this](https://github.com/AntelopeIO/leap/compare/GH-2125-forkdb-simpler...by_best_branch)